Removes escaping of double-quotes around string value

lorengordon opened the pull request:
<!--
1. Please check out and follow our Contributing Guidelines: https://github.com/serverless/serverless/blob/master/CONTRIBUTING.md
2. Do not remove any section of the template. If something is not applicable leave it empty but leave it in the PR
3. Please follow the template, otherwise we'll have to ask you to update it and it will take longer until your PR is merged
-->

## What did you implement:

When using the new string syntax for a variable, a double-quoted string would be rendered with the double-quotes escaped around the value.

## How can we verify it:

serverless.yml:

```
service: hello

provider:
  name: aws
  runtime: nodejs6.10
  environment:
    STRING_DOUBLEQUOTE: ${env:STRING_DOUBLEQUOTE, "foo"}
    STRING_SINGLEQUOTE: ${env:STRING_SINGLEQUOTE, 'foo'}

functions:
  hello:
    handler: handler.hello
```

Rendered cfn template **without** the patch:

```
        "Environment": {
          "Variables": {
            "STRING_DOUBLEQUOTE": "\"foo\"",
            "STRING_SINGLEQUOTE": "foo"
          }
        }
```

Rendered cfn template **with** the patch:

```
        "Environment": {
          "Variables": {
            "STRING_DOUBLEQUOTE": "foo",
            "STRING_SINGLEQUOTE": "foo"
          }
        }
```

<!--
Add any applicable config, commands, screenshots or other resources
to make it easy for us to verify this works. The easier you make it for us
to review a PR, the faster we can review and merge it.

Examples:
* serverless.yml - Fully functioning to easily deploy changes
* Screenshots - Showing the difference between your output and the master
* Cloud Configuration - List cloud resources and show that the correct configuration is in place (e.g. AWS CLI commands)
* Other - Anything else that comes to mind to help us evaluate
-->

## Todos:

- [ ] Write tests
- [ ] Write documentation
- [ ] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

lorengordon commented:
Seeing as the tests pass with and without this patch, I'm pretty sure the tests are not really setting up the inputs and outputs correctly. The tests _look_ reasonable and I'm not seeing what's wrong with them, but if someone wants to point me at the problem, I can patch those too.

lorengordon commented:
Oh I see it. Umm, I think _a lot_ of the variable tests may not be properly setting up the inputs and outputs. The tests are actually mocking the return value of `populateVariable`, which is the value that is being tested.

In effect, we're declaring that `populateVariable` returns `the sky is green` and then testing that the returned value is `the sky is green`. Well, yeah, the logic is circular.

Should probably not be mocking (stubbing?) `populateVariable`, at least not in the overwrite tests.

lorengordon commented:
Removed the stub for `populateVariable`, and now I'm seeing that the `overwrite` stub is also causing the same behavior. That's not right either, we aren't really exercising the overwrite method.

But if I remove the `overwrite` stub, then we get the error: `TypeError: Cannot read property 'stage' of undefined`. Seems to indicate the test setup is not quite right.

lorengordon commented:
Ok, this test exposes the problem on the master branch:

```
    it('should allow a double-quoted string if overwrite syntax provided', () => {
      const serverless = new Serverless();
      const property = 'my stage is ${opt:stage, "prod"}';
      serverless.variables.options = {
        stage: undefined,
      };

      serverless.variables.loadVariableSyntax();

      return serverless.variables.populateProperty(property).then(newProperty => {
        expect(newProperty).to.equal('my stage is prod');

        serverless.variables.overwrite.restore();
        serverless.variables.populateVariable.restore();
        return BbPromise.resolve();
      });
    });
```

lorengordon commented:
There we go. Now the string syntax tests are calling the functions being tested.

HyperBrain commented:
I would check the populate call for a fulfillment just for the sake of completeness and verification that the `newProperty` result can be tested:
```
return expect(serverless.variables.populateProperty(property)).to.be.fulfilled
  .then(newProperty => ...
```

lorengordon commented:
@HyperBrain I just tried that but get, `TypeError: Cannot read property 'then' of undefined`. The setup below does seem to work though. Does this do what you had in mind?

```
      return expect(serverless.variables.populateProperty(property).then(newProperty => {
        expect(newProperty).to.equal('my stage is prod');

        return BbPromise.resolve();
      })).to.be.fulfilled;
```

HyperBrain commented:
@lorengordon Can you check if at the top of the file there is a `chai.use(require('chai-as-promised'));`?

This enables promise handling and testing, exactly as from my example above. The trick is, that you validate the promise, before you "then" it. "fulfilled" is only available when chai-as-promised is loaded as chai plugin.

lorengordon commented:
This is the only line that contains `chai`:

```
const expect = require('chai').expect;
```

HyperBrain commented:
Ok, then replace it with:
```
const chai = require('chai');

chai.use(require('chai-as-promised'));
const expect = chai.expect;
```
and we should have full chai promise support afterwards.
If there are sinon tests too in the unit tests, you cann add `chai.use(require('sinon-chai'));` too, to have the sinon extensions in place too. Then you can do things like `expect(myStub).to.have.been.calledOnce`.

lorengordon commented:
Ok, this diff works using the approach you proposed:

```diff
diff --git a/lib/classes/Variables.test.js b/lib/classes/Variables.test.js
index 19943db7..10286531 100644
--- a/lib/classes/Variables.test.js
+++ b/lib/classes/Variables.test.js
@@ -3,7 +3,7 @@
 const path = require('path');
 const proxyquire = require('proxyquire');
 const YAML = require('js-yaml');
-const expect = require('chai').expect;
+const chai = require('chai');
 const Variables = require('../../lib/classes/Variables');
 const Utils = require('../../lib/classes/Utils');
 const Serverless = require('../../lib/Serverless');
@@ -14,6 +14,10 @@ const AwsProvider = require('../plugins/aws/provider/awsProvider');
 const BbPromise = require('bluebird');
 const os = require('os');

+chai.use(require('chai-as-promised'));
+
+const expect = chai.expect;
+
 describe('Variables', () => {
   describe('#constructor()', () => {
     const serverless = new Serverless();
@@ -182,11 +186,12 @@ describe('Variables', () => {

       serverless.variables.loadVariableSyntax();

-      return serverless.variables.populateProperty(property).then(newProperty => {
-        expect(newProperty).to.equal('my stage is prod');
+      return expect(serverless.variables.populateProperty(property)).to.be.fulfilled
+        .then(newProperty => {
+          expect(newProperty).to.equal('my stage is prod');

-        return BbPromise.resolve();
-      });
+          return BbPromise.resolve();
+        });
     });

     it('should call getValueFromSource if no overwrite syntax provided', () => {
```

HyperBrain commented:
You can even dismiss the `return BbPromise.resolve()` if you just return the `expect()` because chai-as-promised turns the expectation result into promises:
```
 .then(newProperty => expect(newProperty).to.equal('my stage is prod'));
```
Sorry for the lots of single small comments from my side 😃

lorengordon commented:
@HyperBrain I did remove that in the actual commit, and I modified the diff in the comment after the fact...

And no problem on the comments, I appreciate the pointers! Very new to javascript and much of this promise stuff escapes me!

eahefnawy commented:
@lorengordon thanks for that! I see that you've modified some of our tests, but I think it's best if we add a separate test case that specifically tests this edge case. Imo it'd be better to keep our tests focused and high in quality.

Other than that, I've tried your reproduction steps, and they work as expected. 👍

lorengordon commented:
@eahefnawy The tests I modified were invalid. They didn't actually test the primary condition they claimed to test. One of the tests already claimed to be checking for _exactly_ what this patch fixes. I fixed the invalid tests. Now they actually do what they describe.

Please take another look at the modified tests and the discussion above.

Also, I'm afraid several other tests in this module may suffer from the same circular logic. 😢
